README.md
# woocommerce-product-addons
This repository is a paste of the original code woocommerce-product-addon

## Description
WooCommerce Product Add-ons lets you add extra options to products which the user can select. Add-ons can be checkboxes, a select box, or custom input. Each option can optionally be given a price which is added to the cost of the product.

## Official url
https://woocommerce.com/

https://woocommerce.com/products/product-add-ons/


changelog.txt
*** Product Addons Changelog ***

2017-04-03 - version 2.8.1
* Fix - Update for WooCommerce 3.0 compatibility.

2017-02-23 - version 2.8.0
* Fix - When re-adding items to cart, similar named addons is missing.

2016-12-09 - version 2.7.26
* Fix - Issue where required fields are failing to add to the cart using the order again option.
* Fix - Issue product add-ons are not being re-added when paying for pending or failed renewal orders.
* Fix - Update deprecated API of WooCommerce being used this extension.

2016-10-26 - version 2.7.25
* Fix - Text field type addon is not accounting for multibyte characters.

2016-09-29 - version 2.7.24
* Fix - Undefined error when variation data is not found.
* Fix - Product totals not updated instantly when clicking up and down number input.
* Fix - In Safari browser, validation not working when minimum set to 0.

2016-09-06 - version 2.7.23
* Fix - Addon totals were displayed on top of options due to last version.

2016-08-10 - version 2.7.22
* Tweak - Relocate addon fields to be after the variation description.

2016-07-16 - version 2.7.21
* Fix - Tax inclusive display is not showing on addons in cart/checkout.

2016-07-07 - version 2.7.20
* Tweak - Load frontend CSS with wp_enqueue_script hook instead of get_header

2016-06-14 - version 2.7.19
* Fix - Product addons not applying to grouped products when using global addons.

2016-05-26 - version 2.7.18
* Tweak - Adjusts the 'Product Data' icon to use Dashicons
* Fix   - Ensure compatibility with bookings where the price was not affected by the cost multiplier.

2016-04-04 - version 2.7.17
* Fix - Options total/grand total does not work with "Free" product.
* Fix - Missing minified quickview.js.

2016-03-20 - version 2.7.16
* Fix - When fields are not required and left empty, prevented add to cart from working.
* Fix - Compatibility with Subscriptions title when addon is present.
* Tweak - Compatibility with Composite Products.
* Tweak - Clear addon fields when clear variations link is clicked.
* Tweak - Using custom multiplier if zero is entered, don't show addons in cart.
* Update - POT file.
* Remove - accounting.js file in favor of using the one that comes with WC core.

2016-01-26 - version 2.7.15
* Tweak - Show the price display suffix on the grand total.

2016-01-14 - version 2.7.14
* Fix - WooCommerce 2.5 compatibility.
* Fix - Made the 'required' tooltip translatable.

2015-11-24 - version 2.7.13
* Fix - Compatibility fix for WooCommerce Photography. The "select options" button should now lead to the correct screen when using WC Photography 1.0.6.
* i18n / translation updates

2015-11-06 - version 2.7.12
* Fix - Compatibility fix for WooCommerce Photography. Users can no longer add photography products to cart without filling any required add-on fields.
* Fix - The "default label" place holder now clears when you click into it.
* Fix - Fix toggle display on "global add-ons" page.
* Fix - Compatibility fix for WooCommerce Quick View. Prices with variations now quickly display the correct grand total on single product pages.

2015-09-29 - version 2.7.11
* Limit input element names in HTML to less than 64 characters to avoid issues on some hosts

2015-08-20 - version 2.7.10
* Automatically add a default option when a new add-on group is added to a product
* Prevent merchant from removing all options in a given add-on group
* Add custom email as an add-on field option
* Add custom letters only text, digits only text and letters and digits only text as add-on field options
* Add subscription period to add-on prices and totals for simple subscriptions too (not just those with variations)

2015-07-29 - version 2.7.9
* Fix - 2.4 compatibility.
* Tweak - Improved bundles/composite compatibility.
* Tweak - Don't auto select radio buttons.
* Tweak - add filter woocommerce_product_addons_validate_value

2015-07-04 - version 2.7.8
* Fix - Using the POST data to get the product ID caused issues with 3rd party plugins. The issue around grouped products this was added to fix needs to be tackled by a different approach. In this case we decided to..
* Tweak - Make children products inherit addons from the parent (e.g the grouped product parent).
* Tweak - Don't allow addons with required fields to be purchased from grouped product pages.

2015-07-03 - version 2.7.7
* Fix - Compatibility with OPC and other plugins which change the add to cart process.

2015-06-28 - version 2.7.6
* Fix - Only add file upload cost if a file was actually uploaded .

2015-06-19 - version 2.7.5
* Fix - Link product-level add-ons to grouped products when adding to cart
* Fix - Use the correct ID for validation
* Tweak - Allow plugins to hide grand total.

2015-05-27 - version 2.7.4
* Bookings compatibility.

2015-03-13 - version 2.7.3
* Fix duplicates when ordering again.

2015-03-02 - version 2.7.2
* Fix woocommerce_order_item_display_meta_value
* Fix totals calc when currency contains decimal.

2015-02-24 - Version 2.7.1
* Filter upload dir
* Upload attachments linked in the notification emails

2015-01-27 - version 2.7.0
* Renamed main file.
* Make addons script hookable.
* Added WC 2.3.0 compatibility.

2014-12-18 - version 2.6.11
* Reverted previous change

2014-12-12 - version 2.6.10
* Fixed unwanted capitalization on attribute label in order item details of thank you page.

2014-12-02 - version 2.6.9
* Fixed the checkbox add-ons when using re-order action.
* Fixed the file upload path on order confirmation page and order review page.

2014-11-28 - version 2.6.8
* Added hooks to allow Global Add-ons for custom taxonomies.
* Added compressed assets.

2014-09-01 - version 2.6.7
* Fix notices when ordering again.
* Hooks for WPML.

2014-07-21 - version 2.6.6
* input multiplier - Set default value to avoid "This must be a number!".
* Added POT file.
* wc_format_decimal needs to be used on save.
* Correct display after saving global addon.

2014-06-30 - version 2.6.5
* Fixed notices when re-adding item to cart.


woocommerce-product-addons.php
<?php
/*
Plugin Name: WooCommerce Product Add-ons
Plugin URI: https://woocommerce.com/products/product-add-ons/
Description: WooCommerce Product Add-ons lets you add extra options to products which the user can select. Add-ons can be checkboxes, a select box, or custom input. Each option can optionally be given a price which is added to the cost of the product.
Version: 2.8.1
Author: WooCommerce
Author URI: https://woocommerce.com
Requires at least: 3.8
Tested up to: 4.6
WC tested up to: 2.5
	Copyright: © 2009-2017 WooCommerce.
	License: GNU General Public License v3.0
	License URI: http://www.gnu.org/licenses/gpl-3.0.html
*/

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Required functions.
 */
if ( ! function_exists( 'woothemes_queue_update' ) ) {
	require_once( 'woo-includes/woo-functions.php' );
}

/**
 * Plugin updates.
 */
woothemes_queue_update( plugin_basename( __FILE__ ), '147d0077e591e16db9d0d67daeb8c484', '18618' );

if ( is_woocommerce_active() ) {

	/**
	 * Main class.
	 */
	class WC_Product_Addons {

		/**
		 * Constructor.
		 */
		public function __construct() {
			add_action( 'plugins_loaded', array( $this, 'load_plugin_textdomain' ) );
			add_action( 'plugins_loaded', array( $this, 'init_classes' ) );
			add_action( 'init', array( $this, 'init_post_types' ), 20 );
			add_action( 'init', array( $this, 'setup_notices' ) );
			register_activation_hook( __FILE__, array( $this, 'install' ) );
			add_filter( 'plugin_action_links_' . plugin_basename( __FILE__ ), array( $this, 'action_links' ) );
		}

		/**
		 * Localisation.
		 */
		public function load_plugin_textdomain() {
			load_plugin_textdomain( 'woocommerce-product-addons', false, dirname( plugin_basename( __FILE__ ) ) . '/languages/' );
		}

		/**
		 * Initializes plugin classes.
		 */
		public function init_classes() {
			if ( is_admin() ) {
				$this->init_admin();
			}

			include_once( dirname( __FILE__ ) . '/includes/class-product-addon-display.php' );
			include_once( dirname( __FILE__ ) . '/includes/class-product-addon-cart.php' );
			include_once( dirname( __FILE__ ) . '/includes/class-product-addon-ajax.php' );

			// Handle WooCommerce 3.0 compatibility.
			if ( defined( 'WC_VERSION' ) && version_compare( WC_VERSION, '3.0.0', '<' ) ) {
				include_once( dirname( __FILE__ ) . '/includes/legacy/class-product-addon-display-legacy.php' );
				include_once( dirname( __FILE__ ) . '/includes/legacy/class-product-addon-cart-legacy.php' );
				include_once( dirname( __FILE__ ) . '/includes/legacy/class-wc-addons-ajax.php' );

				$GLOBALS['Product_Addon_Display'] = new Product_Addon_Display_Legacy();
				$GLOBALS['Product_Addon_Cart']    = new Product_Addon_Cart_Legacy();
				new WC_Addons_Ajax();
			} else {
				$GLOBALS['Product_Addon_Display'] = new Product_Addon_Display();
				$GLOBALS['Product_Addon_Cart']    = new Product_Addon_Cart();
				new Product_Addon_Cart_Ajax();
			}
		}

		/**
		 * Initializes plugin admin.
		 */
		protected function init_admin() {
			include_once( dirname( __FILE__ ) . '/admin/class-product-addon-admin.php' );

			// Handle WooCommerce 3.0 compatibility.
			if ( defined( 'WC_VERSION' ) && version_compare( WC_VERSION, '3.0.0', '<' ) ) {
				include_once( dirname( __FILE__ ) . '/admin/legacy/class-product-addon-admin-legacy.php' );

				$GLOBALS['Product_Addon_Admin'] = new Product_Addon_Admin_Legacy();
			} else {
				$GLOBALS['Product_Addon_Admin'] = new Product_Addon_Admin();
			}
		}

		/**
		 * Init post types used for addons.
		 */
		public function init_post_types() {
			register_post_type( 'global_product_addon',
				array(
					'public'              => false,
					'show_ui'             => false,
					'capability_type'     => 'product',
					'map_meta_cap'        => true,
					'publicly_queryable'  => false,
					'exclude_from_search' => true,
					'hierarchical'        => false,
					'rewrite'             => false,
					'query_var'           => false,
					'supports'            => array( 'title' ),
					'show_in_nav_menus'   => false,
				)
			);

			register_taxonomy_for_object_type( 'product_cat', 'global_product_addon' );
		}

		/**
		 * Plugin action links
		 */
		public function action_links( $links ) {
			$plugin_links = array(
				'<a href="https://support.woocommerce.com/">' . __( 'Support', 'woocomerce-product-addons' ) . '</a>',
				'<a href="https://docs.woocommerce.com/document/product-add-ons/">' . __( 'Documentation', 'woocomerce-product-addons' ) . '</a>',
			);
			return array_merge( $plugin_links, $links );
		}

		/**
		 * Setup notices.
		 */
		public function setup_notices() {
			add_action( 'admin_notices', array( $this, 'customizer_notice' ) );
		}

		/**
		 * Installation.
		 * Runs on activation. Assigns a notice message to a WordPress option.
		 */
		public function install() {
			$notices   = get_option( 'wpa_activation_notice', array() );
			$notices[] = '<p><strong>' . __( 'Thanks for installing WooCommerce Product Add-ons.', 'woocommerce-product-addons' ) . '</strong></p><p>' . sprintf( __( 'Before diving in, we highly recommend getting to grips with %1$sper product add-ons%2$s and %3$sglobal add-ons%2$s in the %4$sdocumentation%2$s.', 'woocommerce-product-addons' ), '<a href="https://docs.woocommerce.com/document/product-add-ons/#section-1">', '</a>', '<a href="https://docs.woocommerce.com/document/product-add-ons/#section-2">', '<a href="https://docs.woocommerce.com/document/product-add-ons/">' ) . '</p><p>' . sprintf( __( 'Ready to get started? %sAdd a global add-on%s.','woocommerce-product-addons' ), '<a href="' . esc_url( admin_url() ) . 'edit.php?post_type=product&page=global_addons">', '</a>' ) . '</p>';
			update_option( 'wpa_activation_notice', $notices );
		}

		/**
		 * Admin notice.
		 * Checks the notice setup in install(). If it exists display it then delete the option so it's not displayed again.
		 */
		public function customizer_notice() {
			$notices = get_option( 'wpa_activation_notice' );
			if ( $notices = get_option( 'wpa_activation_notice' ) ) {
				foreach ( $notices as $notice ) {
					echo '<div class="notice is-dismissible updated">' . $notice . '</div>';
				}
				delete_option( 'wpa_activation_notice' );
			}
		}
	}

	new WC_Product_Addons();

	/**
	 * Gets addons assigned to a product by ID.
	 *
	 * @param  int    $post_id ID of the product to get addons for.
	 * @param  string $prefix for addon field names. Defaults to postid.
	 * @param  bool   $inc_parent Set to false to not include parent product addons.
	 * @param  bool   $inc_global Set to false to not include global addons.
	 * @return array
	 */
	function get_product_addons( $post_id, $prefix = false, $inc_parent = true, $inc_global = true ) {
		if ( ! $post_id ) {
			return array();
		}

		$addons     = array();
		$raw_addons = array();

		if ( defined( 'WC_VERSION' ) && version_compare( WC_VERSION, '3.0.0', '<' ) ) {
			$product_terms  = apply_filters( 'get_product_addons_product_terms', wp_get_post_terms( $post_id, 'product_cat', array( 'fields' => 'ids' ) ), $post_id );
			$exclude        = get_post_meta( $post_id, '_product_addons_exclude_global', true );
			$product_addons = array_filter( (array) get_post_meta( $post_id, '_product_addons', true ) );
		} else {
			$product        = wc_get_product( $post_id );
			$product_terms  = apply_filters( 'get_product_addons_product_terms', wc_get_object_terms( $product->get_id(), 'product_cat', 'term_id' ), $product->get_id() );
			$exclude        = $product->get_meta( '_product_addons_exclude_global' );
			$product_addons = array_filter( (array) $product->get_meta( '_product_addons' ) );
		}

		// Product Parent Level Addons.
		if ( $inc_parent && $parent_id = wp_get_post_parent_id( $post_id ) ) {
			$raw_addons[10]['parent'] = apply_filters( 'get_parent_product_addons_fields', get_product_addons( $parent_id, $parent_id . '-', false, false ), $post_id, $parent_id );
		}

		// Product Level Addons.
		$raw_addons[10]['product'] = apply_filters( 'get_product_addons_fields', $product_addons, $post_id );

		// Global level addons (all products).
		if ( '1' !== $exclude && $inc_global ) {
			$args = array(
				'posts_per_page'   => -1,
				'orderby'          => 'meta_value',
				'order'            => 'ASC',
				'meta_key'         => '_priority',
				'post_type'        => 'global_product_addon',
				'post_status'      => 'publish',
				'suppress_filters' => true,
				'meta_query' => array(
					array(
						'key'   => '_all_products',
						'value' => '1',
					),
				),
			);

			$global_addons = get_posts( $args );

			if ( $global_addons ) {
				foreach ( $global_addons as $global_addon ) {
					$priority                                     = get_post_meta( $global_addon->ID, '_priority', true );
					$raw_addons[ $priority ][ $global_addon->ID ] = apply_filters( 'get_product_addons_fields', array_filter( (array) get_post_meta( $global_addon->ID, '_product_addons', true ) ), $global_addon->ID );
				}
			}

			// Global level addons (categories).
			if ( $product_terms ) {
				$args = apply_filters( 'get_product_addons_global_query_args', array(
					'posts_per_page'   => -1,
					'orderby'          => 'meta_value',
					'order'            => 'ASC',
					'meta_key'         => '_priority',
					'post_type'        => 'global_product_addon',
					'post_status'      => 'publish',
					'suppress_filters' => true,
					'tax_query'        => array(
						array(
							'taxonomy'         => 'product_cat',
							'field'            => 'id',
							'terms'            => $product_terms,
							'include_children' => false,
						),
					),
				), $product_terms );

				$global_addons = get_posts( $args );

				if ( $global_addons ) {
					foreach ( $global_addons as $global_addon ) {
						$priority                                     = get_post_meta( $global_addon->ID, '_priority', true );
						$raw_addons[ $priority ][ $global_addon->ID ] = apply_filters( 'get_product_addons_fields', array_filter( (array) get_post_meta( $global_addon->ID, '_product_addons', true ) ), $global_addon->ID );
					}
				}
			}
		}

		ksort( $raw_addons );

		foreach ( $raw_addons as $addon_group ) {
			if ( $addon_group ) {
				foreach ( $addon_group as $addon ) {
					$addons = array_merge( $addons, $addon );
				}
			}
		}

		// Generate field names with unqiue prefixes.
		if ( ! $prefix ) {
			$prefix = apply_filters( 'product_addons_field_prefix', "{$post_id}-", $post_id );
		}

		// Let's avoid exceeding the suhosin default input element name limit of 64 characters.
		$max_addon_name_length = 45 - strlen( $prefix );

		// If the product_addons_field_prefix filter results in a very long prefix, then
		// go ahead and enforce sanity, exceed the default suhosin limit, and just use
		// the prefix and the field counter for the input element name.
		if ( $max_addon_name_length < 0 ) {
			$max_addon_name_length = 0;
		}

		$addon_field_counter = 0;

		foreach ( $addons as $addon_key => $addon ) {
			if ( empty( $addon['name'] ) ) {
				unset( $addons[ $addon_key ] );
				continue;
			}
			if ( empty( $addons[ $addon_key ]['field-name'] ) ) {
				$addon_name = substr( $addon['name'], 0, $max_addon_name_length );
				$addons[ $addon_key ]['field-name'] = sanitize_title( $prefix . $addon_name . '-' . $addon_field_counter );
				$addon_field_counter++;
			}
		}

		return apply_filters( 'get_product_addons', $addons );
	}

	/**
	 * Display prices according to shop settings.
	 *
	 * @param  float      $price     Price to display.
	 * @param  WC_Product $cart_item Product from cart.
	 * @return float
	 */
	function get_product_addon_price_for_display( $price, $cart_item = null ) {
		global $product;

		if ( '' === $price || '0' == $price ) {
			return;
		}

		if ( ( is_cart() || is_checkout() ) && null !== $cart_item ) {
			// Support new WooCommerce 3.0 WC_Product->get_id().
			if ( is_callable( $cart_item, 'get_id' ) ) {
				$product = wc_get_product( $cart_item->get_id() );
			} else {
				$product = wc_get_product( $cart_item->id );
			}
		}

		if ( is_object( $product ) ) {
			// Support new wc_get_price_excluding_tax() and wc_get_price_excluding_tax() functions.
			if ( function_exists( 'wc_get_price_excluding_tax' ) ) {
				$display_price = get_product_addon_tax_display_mode() === 'incl' ? wc_get_price_including_tax( $product, array( 'qty' => 1, 'price' => $price ) ) : wc_get_price_excluding_tax( $product, array( 'qty' => 1, 'price' => $price ) );
			} else {
				$display_price = get_product_addon_tax_display_mode() === 'incl' ? $product->get_price_including_tax( 1, $price ) : $product->get_price_excluding_tax( 1, $price );
			}
		} else {
			$display_price = $price;
		}

		return $display_price;
	}

	/**
	 * Return tax display mode depending on context.
	 *
	 * @return string
	 */
	function get_product_addon_tax_display_mode() {
		if ( is_cart() || is_checkout() ) {
			return get_option( 'woocommerce_tax_display_cart' );
		}

		return get_option( 'woocommerce_tax_display_shop' );
	}
}
